### PR TITLE
Hash expressions

### PR DESCRIPTION
**Pull request by lbartoletti**
## Description

This PR adds hash functions to QGIS. No new dependency is required since it uses [QCryptographicHash](https://doc.qt.io/qt-5/qcryptographichash.html) available in QtCore.

Use case is the same as PgCrypto but in client side for all supported format.
 
## Checklist

- [X] Commit messages are descriptive and explain the rationale for changes
- [ ] Commits which fix bugs include `Fixes #11111` at the bottom of the commit message
- [X] I have read the [QGIS Coding Standards](https://docs.qgis.org/testing/en/docs/developers_guide/codingstandards.html) and this PR complies with them
- [X] New unit tests have been added for core changes
- [X] I have run [the `scripts/prepare-commit.sh` script](https://github.com/qgis/QGIS/blob/master/.github/CONTRIBUTING.md#contributing-to-qgis) before each commit
- [ ] I have evaluated whether it is appropriate for this PR to be backported, backport requests are left as label or comment

### CODE REVIEW

**Review by nyalldawson — APPROVED** [review]
Just gotta fix the build ;)

**Comment by lbartoletti** [comment]
> hash(value, algorithm='keccak_224')

:+1: 

I would like to keep also separated functions since I often use md5(value) / sha256(value) etc. :)
Do you agree?

**Comment by Gustry** [comment]
IMHO, if it's a single function, it may fit in the already existing group `Conversions` with `to_hash` (same pattern as existing QGIS expressions) ?

**Comment by m-kuhn** [comment]
I can see the use for shorthands for sha256 and md5, but I wonder if this necessarily means that we have to provide all of them. No strong opinion.
Concerning the method naming, I'd stick to `hash`, hasing is a one-way operation (in contrast to the usual to_xxx cast conversions)

**Comment by NathanW2** [comment]
Yeah I think just a single hash function makes the most sense. Avoids
adding more functions then are needed

**Comment by lbartoletti** [comment]
So, only one function `hash(value, string)`. Dedicated group or existing group (which one : `Conversions`? )?

Thanks

**Comment by m-kuhn** [comment]
How about
* `hash(value, method)`
* `md5(value)`
* `sha256(value)`
* and `keccak256(value)`?

in `Conversions`?

**Comment by lbartoletti** [comment]
@m-kuhn done. Good for you?

File: src/core/expression/qgsexpressionfunction.cpp [hunk 1]
```diff
@@ -4991,8 +4991,90 @@ static QVariant fcnFileSize( const QVariantList &values, const QgsExpressionCont
...
+  return QString( QCryptographicHash::hash( str.toUtf8(), algorithm ).toHex() );
+}
 
+static QVariant fcnGenericHash( const QVariantList &values, const QgsExpressionContext *, QgsExpression *parent, const QgsExpressionNodeFunction * )
+{
+  QVariant hash;
+  QString str = QgsExpressionUtils::getStringValue( values.at( 0 ), parent );
+  QString method = QgsExpressionUtils::getStringValue( values.at( 1 ), parent ).toLower();
+
+  if ( method == QString( "md4" ) )
```
Comment (m-kuhn) [inline src/core/expression/qgsexpressionfunction.cpp]:
```suggestion
  if ( method == QLatin1String( "md4" ) )
```

**Comment by m-kuhn** [comment]
Yep, pretty much, thanks :+1:

**Review by m-kuhn — APPROVED** [review]

**Pull request merged**